### PR TITLE
Give a `Conv2D` layer call an output shape

### PR DESCRIPTION
**Pull request by khatchad**
Fixes wala/ML#820.

A `Conv2D` layer call produced no output shape, so every value downstream of it lost its rank. The dtype survived; only the shape was lost.

## The Cause

The model file described `conv2d`, the function reachable as `tf.nn.conv2d`. Nothing described `tf.keras.layers.Conv2D`, the layer class whose instance is called. Layer calls are modeled selectively, with generators for `MaxPool` and `GlobalAveragePooling1D`, so this was a hole in that family rather than an absent capability.

## The Fix

Modeled the way `Dense` is, in three parts: a constructor keeping the argument that fixes the output's last axis, a callable allocating a fresh tensor, and a generator computing the result.

Both layers read that stored argument the same way and differ only in its name, `units` against `filters`, so the lookup became an overridable hook rather than a second copy of it.

## Why The Spatial Extents Are Unresolved

A convolution keeps the batch axis, rewrites the two spatial axes, and sets the channel axis to `filters`. The spatial extents are fixed integers at run time, determined by the input extent together with `kernel_size`, `strides`, `padding` and `dilation_rate`. Folding that arithmetic is separate work, and per the dimension conventions an extent the analysis cannot compute but which is not feed-dependent is `Unresolved` rather than `Dynamic`: nothing here carries evidence that the runtime shape would report `None`.

Recovering the rank is what unblocks the chain. `Dense` rewrites only the last axis, so a following `Flatten` and `Dense` reach the correct final shape from a rank-4 input whatever the spatial extents are, which is the pattern this distils:

```python
pred = conv_net(x)          # rank recovered here
loss = cross_entropy_loss(pred, y)
acc  = accuracy(pred, y)
```

## Testing

The reproduction committed with the issue flips from a known failure to a positive guard, and its control, the identical chain without the convolution, still passes.

One existing pin moves for the right reason. The test `testEagerExecution` sees its tensor-local count rise from four to five: that fixture builds a `Conv2D` and calls it, so its result now counts as tensor-typed where before it carried nothing at all.

## Scope

`MaxPool2D` is absent from the model file for the same reason and is untouched here. Whether the remaining convolution and pooling layer classes need the same treatment is worth a sweep rather than another single addition.

### CODE REVIEW

**Comment by khatchad** [comment]
Self-review, since the automated reviewer hit its quota.

**One placement decision worth stating rather than leaving implicit.** New operations are supposed to register in the shared dispatch chain, which serves both anchorings from a single arm. This one does not: it sits beside the `Dense` arm it mirrors, at line 1795, well past the shared chain, and is registered separately in the manual walker.

That is deliberate. `Conv2DCall` extends `DenseCall` and differs only in the name of the constructor argument it reads back, so the two dispatch together or the pair is split across two mechanisms for no gain. Registering in both tables satisfies the invariant the shared chain exists to enforce, and the coverage check passes. Moving the pair is worth doing together, not one half of it.

**What the fix does not do.** The spatial extents are unresolved rather than computed. That is a real limitation and the reason the reproduction asserts only the final shape: `Dense` rewrites the last axis, so the chain reaches `(128, 10)` from a rank-4 input whatever the middle holds. Folding the `kernel_size`, `strides`, `padding` and `dilation_rate` arithmetic would make the intermediate shapes concrete and is separate work.

**Verified rather than assumed.** Clean build from `mvn clean install`, since the model file changed and its resources are cached: 1202 tests, 0 failures, 3 skipped. The reproduction flips from a known failure to a positive guard, its control still passes, and one existing pin moves for the right reason, with the reason recorded in its Javadoc.
